### LAB TASK III
### NAME: MUHAMMAD HARIS
### REGISTRATION. NO: 20MDBCS087
### SUBJECT: DATA ANALYSIS LAB
### SUBMITTED TO: DR. TARIQ SADAD

<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="#Data-Acquisition">Data Acquisition</a>
    <li><a href="#Basic-Insights-from-the-Data-set">Basic Insights from the Data set</a></li>
</ol>


</div>
<hr>


# Data Acquisition
<p>
A data set is typically a file containing data stored in one of several formats. Common file formats containing data sets include: .csv, .json, .xlsx etc. The data set can be stored in different places, on your local machine, on a server or a websiite, cloud storage and so on.<br>

To analyse data in a Python notebook, we need to bring the data set into the notebook. In this section, you will learn how to load a data set into our Jupyter Notebook.<br>

In our case, the Automobile Data set is an online source, and it is in a CSV (comma separated value) format. Let's use this data set as an example to practice reading data.
<ul>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a very popular and very useful tool that enables us to read various datasets into a data frame; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


In [ ]:
# uncomment the lines below if you need to install specific version of libraries if using this notebook
# in an environment where these libraries are not installed
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [1]:
# import pandas library
import pandas as pd
import numpy as np

<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage">HERE</a> for free storage.


The functions below will download the dataset into your browser:


In [2]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

In [16]:
file_path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv'

To obtain the dataset, utilize the download() function as defined above:


In [17]:
await download(file_path, "auto.csv")
file_name="auto.csv"

Utilize the Pandas method `read_csv()` to load the data into a dataframe.


In [21]:
df = pd.read_csv('./auto.csv')

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [ ]:
#filepath = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"
#df = pd.read_csv(filepath, header=None)

After reading the data set, we can use the <code>data_frame.head(n)</code> method to check the top n rows of the data frame, where n is an integer. Contrary to <code>data_frame.head(n)</code>, <code>data_frame.tail(n)</code> will show you the bottom n rows of the data frame.


In [22]:
df = pd.read_csv('./auto.csv')
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe")
df.head(5)
df.shape

The first 5 rows of the dataframe


(204, 26)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [23]:
# Write your code below and press Shift+Enter to execute
print("The last 10 rows of the dataframe\n")
df.tail(10)

The last 10 rows of the dataframe



3    ? alfa-romero     gas    std   two convertible  rwd  front   88.6  \
194 -1   74       volvo     gas    std  four       wagon  rwd  front  104.3   
195 -2  103       volvo     gas    std  four       sedan  rwd  front  104.3   
196 -1   74       volvo     gas    std  four       wagon  rwd  front  104.3   
197 -2  103       volvo     gas  turbo  four       sedan  rwd  front  104.3   
198 -1   74       volvo     gas  turbo  four       wagon  rwd  front  104.3   
199 -1   95       volvo     gas    std  four       sedan  rwd  front  109.1   
200 -1   95       volvo     gas  turbo  four       sedan  rwd  front  109.1   
201 -1   95       volvo     gas    std  four       sedan  rwd  front  109.1   
202 -1   95       volvo  diesel  turbo  four       sedan  rwd  front  109.1   
203 -1   95       volvo     gas  turbo  four       sedan  rwd  front  109.1   

     ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
194  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  13415  
195  ...  141  mpfi  3.78  3.15   9.5  114  5400  24  28  15985  
196  ...  141  mpfi  3.78  3.15   9.5  114  5400  24  28  16515  
197  ...  130  mpfi  3.62  3.15   7.5  162  5100  17  22  18420  
198  ...  130  mpfi  3.62  3.15   7.5  162  5100  17  22  18950  
199  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
200  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
201  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
202  ...  145   idi  3.01  3.40  23.0  106  4800  26  27  22470  
203  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[10 rows x 26 columns]

<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at the data set. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe the data, you can introduce a header. This information is available at:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Thus, you have to add headers manually.
</p>
<p>
First, create a list "headers" that include all column names in order.
Then, use <code>dataframe.columns = headers</code> to replace the headers with the list you created.
</p>


In [24]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


 Replace headers and recheck our data frame:


In [25]:
df.columns = headers
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

You can also see the first 10 entries of the updated data frame and note that the headers are updated.


In [26]:
df.head(10)
df_age = df['engine-type']
df_age

0      dohc
1      ohcv
2       ohc
3       ohc
4       ohc
       ... 
199     ohc
200     ohc
201    ohcv
202     ohc
203     ohc
Name: engine-type, Length: 204, dtype: object

Now, we need to replace the "?" symbol with NaN so the dropna() can remove the missing values:


In [27]:
#df1=df.replace('?',np.NaN)
df_sub = df[['engine-type', 'price']]

df_sub.head()
df
df.loc[11]

symboling                0
normalized-losses      188
make                   bmw
fuel-type              gas
aspiration             std
num-of-doors           two
body-style           sedan
drive-wheels           rwd
engine-location      front
wheel-base           101.2
length               176.8
width                 64.8
height                54.3
curb-weight           2710
engine-type            ohc
num-of-cylinders       six
engine-size            164
fuel-system           mpfi
bore                  3.31
stroke                3.19
compression-ratio      9.0
horsepower             121
peak-rpm              4250
city-mpg                21
highway-mpg             28
price                20970
Name: 11, dtype: object

You can drop missing values along the column "price" as follows:


In [29]:
df=df.dropna(subset=["price"], axis=0)
df.head(20)

symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           1                 ?  alfa-romero       gas        std   
2           2               164         audi       gas        std   
3           2               164         audi       gas        std   
4           2                 ?         audi       gas        std   
5           1               158         audi       gas        std   
6           1                 ?         audi       gas        std   
7           1               158         audi       gas      turbo   
8           0                 ?         audi       gas      turbo   
9           2               192          bmw       gas        std   
10          0               192          bmw       gas        std   
11          0               188          bmw       gas        std   
12          0               188          bmw       gas        std   
13          1                 ?          bmw       gas        std   
14          0                 ?          bmw       gas        std   
15          0                 ?          bmw       gas        std   
16          0                 ?          bmw       gas        std   
17          2               121    chevrolet       gas        std   
18          1                98    chevrolet       gas        std   
19          0                81    chevrolet       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two    hatchback          rwd           front        94.5  ...   
2          four        sedan          fwd           front        99.8  ...   
3          four        sedan          4wd           front        99.4  ...   
4           two        sedan          fwd           front        99.8  ...   
5          four        sedan          fwd           front       105.8  ...   
6          four        wagon          fwd           front       105.8  ...   
7          four        sedan          fwd           front       105.8  ...   
8           two    hatchback          4wd           front        99.5  ...   
9           two        sedan          rwd           front       101.2  ...   
10         four        sedan          rwd           front       101.2  ...   
11          two        sedan          rwd           front       101.2  ...   
12         four        sedan          rwd           front       101.2  ...   
13         four        sedan          rwd           front       103.5  ...   
14         four        sedan          rwd           front       103.5  ...   
15          two        sedan          rwd           front       103.5  ...   
16         four        sedan          rwd           front       110.0  ...   
17          two    hatchback          fwd           front        88.4  ...   
18          two    hatchback          fwd           front        94.5  ...   
19         four        sedan          fwd           front        94.5  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           152         mpfi  2.68    3.47               9.0        154   
2           109         mpfi  3.19    3.40              10.0        102   
3           136         mpfi  3.19    3.40               8.0        115   
4           136         mpfi  3.19    3.40               8.5        110   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           131         mpfi  3.13    3.40               8.3        140   
8           131         mpfi  3.13    3.40               7.0        160   
9           108         mpfi  3.50    2.80               8.8        101   
10          108         mpfi  3.50    2.80               8.8        101   
11          164

Here, `axis=0` means that the contents along the entire row will be dropped wherever the entity 'price' is found to be NaN

Now, you have successfully read the raw data set and added the correct headers into the data frame.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [30]:
# Write your code below and press Shift+Enter to execute
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables you to save the data set to CSV. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you save the data frame <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


You can also read and save other file formats. You can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


# Basic Insights from the Data set
<p>
After reading data into Pandas dataframe, it is time for you to explore the data set.<br>

There are several ways to obtain essential insights of the data to help you better understand it.
</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas data frames are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, you should always know the data type of each column. In Pandas:
</p>


In [31]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Returns a series with the data type of each column.


In [32]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


<p>
As shown above, you can clearly to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; you will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column such as count, column mean value, column standard deviation, etc., use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [33]:
df.describe()

symboling  wheel-base      length       width      height  \
count  204.000000  204.000000  204.000000  204.000000  204.000000   
mean     0.823529   98.806373  174.075000   65.916667   53.749020   
std      1.239035    5.994144   12.362123    2.146716    2.424901   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.075000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.200000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   204.000000   204.000000         204.000000  204.000000   204.000000  
mean   2555.602941   126.892157          10.148137   25.240196    30.769608  
std     521.960820    41.744569           3.981000    6.551513     6.898337  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.575000   19.000000    25.000000  
50%    2414.000000   119.500000           9.000000   24.000000    30.000000  
75%    2939.250000   142.000000           9.400000   30.000000    34.500000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.
<br>

However, what if you would also like to check all the columns including those that are of type object?
<br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Try it again.

</p>


In [34]:
# describe all the columns in "df"
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   204.000000               204     204       204        204   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                40      32       184        167   
mean      0.823529               NaN     NaN       NaN        NaN   
std       1.239035               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           204        204          204             204  204.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             201         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.806373  ...   
std             NaN        NaN          NaN             NaN    5.994144  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    204.000000          204   204     204        204.000000        204   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           93    23      20               NaN         19   
mean     126.892157          NaN   NaN     NaN         10.148137        NaN   
std       41.744569          NaN   NaN     NaN          3.981000        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.575000        NaN   
50%      119.500000          NaN   NaN     NaN          9.000000        NaN   
75%      142.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        204  204.000000  204.000000   204  
unique        24         NaN         NaN   186  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.240196   30.769608   NaN  
std          NaN    6.551513    6.898337   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.500000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

YOu can now see how many unique values there, which one is the top value, and the frequency of the top value in the object-typed columns.<br>

Some values in the table above show "NaN". Those numbers are not available regarding a particular column type.<br>
</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [35]:
# Write your code below and press Shift+Enter to execute
df[['length','compression-ratio']].describe()

length  compression-ratio
count  204.000000         204.000000
mean   174.075000          10.148137
std     12.362123           3.981000
min    141.100000           7.000000
25%    166.300000           8.575000
50%    173.200000           9.000000
75%    183.200000           9.400000
max    208.100000          23.000000


<details><summary>Click here for the solution</summary>

```python
df[['length', 'compression-ratio']].describe()
```

</details>


<h2>Info</h2>
You can also use another method to check your data set:


It provides a concise summary of your data frame.

This method prints information about a data frame including the index dtype and columns, non-null values and memory usage.


In [36]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 204 entries, 0 to 203
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          204 non-null    int64  
 1   normalized-losses  204 non-null    object 
 2   make               204 non-null    object 
 3   fuel-type          204 non-null    object 
 4   aspiration         204 non-null    object 
 5   num-of-doors       204 non-null    object 
 6   body-style         204 non-null    object 
 7   drive-wheels       204 non-null    object 
 8   engine-location    204 non-null    object 
 9   wheel-base         204 non-null    float64
 10  length             204 non-null    float64
 11  width              204 non-null    float64
 12  height             204 non-null    float64
 13  curb-weight        204 non-null    int64  
 14  engine-type        204 non-null    object 
 15  num-of-cylinders   204 non-null    object 
 16  engine-size        204 non

<h1>Excellent! You have just completed the  Introduction Notebook.</h1>


### Thank you for completing this lab.

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/" target="_blank">Fiorella Wenver</a>

<a href=" https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

<a href="https://www.coursera.org/instructor/~129186572/ " target="_blank" >Abhishek Gagneja</a>
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                       |
| ----------------- | ------- | ---------- | ---------------------------------------- |
| 2023-09-28        | 2.4     | Abhishek Gagneja| Minor instructional update          |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                   |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'     |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab       |


<hr>

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
